[WIP] Use framework provided error class and stack trace as error message

roywei opened the pull request:
*Issue #, if available:*

*Description of changes:*

Each SM library (e.g. SMData Parallel) can passing custom error classes for training toolkit to monitor. It now can show up in SM console as failure reason, and provide a de-duplicated stack trace as error message. 

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-training-toolkit/blob/master/CONTRIBUTING.md) doc
- [ ] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-training-toolkit/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have used the regional endpoint when creating S3 and/or STS clients (if appropriate)
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-training-toolkit/blob/master/README.md)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

satishpasumarthi commented on src/sagemaker_training/process.py:
```diff
@@ -201,6 +209,12 @@ def check_error(cmd, error_class, processes_per_host, cwd=None, capture_error=Tr
             extra_info = "OutOfMemory: Process killed by SIGKILL (signal 9)"
+        error_class = errors.ExecuteUserScriptError
+        for error_name in error_classes:
+            if error_name in stderr:
```
Will there be a scenario where there be multiple errors reported. Example: NCCL related error eventually causing mpi to fail (MPI related error) ? If so, are we settling for the first error that is seen in the `error_classes` ?

roywei replied:
we are only interested in specific error that will attribute a Deep Engine library caused this error. So we will not scan for general MPI ABORT or NCCL error. Error names in error_classes as all targeted library specific errors and usually only 1 type of error in log. Even if multiple are found, only 1 is enough for error attribution.(Blame the library for causing the error). So we are good here.

satishpasumarthi commented on src/sagemaker_training/process.py:
```diff
@@ -177,20 +183,22 @@ def check_error(cmd, error_class, processes_per_host, cwd=None, capture_error=Tr
             **kwargs,
         )
-        stderr = output[1]
+        stderr = " ".join(output)
```
output is a list of size 2. The `output[0]` has stdout and `output[1]` has stderr.  Why are we combining both here? The `stdout` might contain normal log messages as well. Could you please clarify?

satishpasumarthi commented on src/sagemaker_training/process.py:
```diff
@@ -177,20 +183,22 @@ def check_error(cmd, error_class, processes_per_host, cwd=None, capture_error=Tr
             **kwargs,
         )
-        stderr = output[1]
+        stderr = " ".join(output)
```
output is a list of size 2. The `output[0]` has stdout and `output[1]` has stderr.  Why are we combining both here? The `stdout` might contain normal log messages as well. Could you please clarify?

roywei replied:
output is the combination of watch function over stdout and stderr, and I think watch only add contents to output is an error is detected.
```
output = await asyncio.gather(
        watch(proc.stdout, error_classes, processes_per_host), watch(proc.stderr, error_classes, processes_per_host)
    )
```
so output should only contain error messages from both stderr and stdout, and it's mostly empty before this PR.

Normal processed logs are printed out here?
https://github.com/aws/sagemaker-training-toolkit/blob/master/src/sagemaker_training/process.py#L82

satishpasumarthi reviewed (commented):
Left few comments !

roywei commented:
Hi @satishpasumarthi @josephevans Could you help take a look at the PR again? Thanks!

We addressed comments and added both SMDDP and SMMP errors tracking support.

roywei commented:
@josephevans @satishpasumarthi all tests are passing now.

satishpasumarthi commented on src/sagemaker_training/process.py:
```diff
@@ -80,29 +100,41 @@ async def watch(stream, proc_per_host):
+async def run_async(cmd, processes_per_host, env, cwd, stderr, error_classes=None, **kwargs):
     """Method responsible for launching asyncio subprocess shell
-    Use asyncio gather to collect processed stdout and stderr
+    Usyncse asyncio gather to collect processed stdout and stderr
```
nit: typo. should be `Use`

satishpasumarthi commented on src/sagemaker_training/mpi.py:
```diff
@@ -23,11 +23,29 @@
+    # list of exceptions SMMP wants training toolkit to catch and log
+    exception_classes = [x for x in dir(exceptions) if isclass(getattr(exceptions, x))]
+except ImportError:
+    logger.info("No exception classes found in smdistributed.modelparallel")
```
Please update the package name to `smdistributed.modelparallel.backend` to distinguish from below.

satishpasumarthi commented on src/sagemaker_training/mpi.py:
```diff
@@ -23,11 +23,29 @@
+    # list of torch exceptions SMMP wants training toolkit to catch and log
+    exception_classes += [x for x in dir(torch_exceptions) if isclass(getattr(torch_exceptions, x))]
+except ImportError:
+    logger.info("No torch exception classes found in smdistributed.modelparallel")
```
Please update the package name to `smdistributed.modelparallel.torch`

satishpasumarthi commented on src/sagemaker_training/mpi.py:
```diff
@@ -23,11 +23,29 @@
+except ImportError:
+    logger.info("No torch exception classes found in smdistributed.modelparallel")
+if not exception_classes:
+    exception_classes = [errors.ExecuteUserScriptError]
```
Could you remove this if condition and add before the `try` statements ? so we can remove assigning empty list in first `except` case and reassigning here.

satishpasumarthi commented on src/sagemaker_training/process.py:
```diff
@@ -111,26 +143,35 @@ async def run_async(cmd, processes_per_host, env, cwd, stderr, **kwargs):
     Args:
         cmd (list): The command to be run.
-        error_class (cls): The class to use when raising an exception.
+        error_classes (list): List of exception classes to watch and raise.
```
Looks like this is an optional list argument? When modelparallel is not enabled (in `mpi.py`, this is still a class and you are wrapping it into list on the fly in the watch function.

roywei replied:
Yes, it's for backward compatibility. I don't want to break any existing code, so made it optional list, also passing a single class would also work.

satishpasumarthi commented on src/sagemaker_training/mpi.py:
```diff
@@ -23,11 +23,29 @@
+except ImportError:
+    logger.info("No exception classes found in smdistributed.modelparallel")
+    exception_classes = []
+try:
```
can we leave an empty line for readability reasons between multiple `try` `except` blocks?

satishpasumarthi reviewed (changes requested):
Left few comments.

satishpasumarthi reviewed (approved):
The changes LGTM.
